### PR TITLE
Belongs To Many targetBindingKey proposal

### PR DESCRIPTION
**Pull request by inoas**
Ref: https://github.com/cakephp/cakephp/issues/10288

### CODE REVIEW

File: src/ORM/Association/BelongsToMany.php [hunk 1]
```diff
@@ -157,6 +164,54 @@ class BelongsToMany extends Association
...
+
+    /**
+     * Gets the name of the field representing the binding key to the target table.
+     *
+     * @return string
+     */
+    public function getTargetBindingKey()
+    {
+        if ($this->_targetBindingKey === null) {
+            if ($this->_targetBindingKey === null) {
```
Comment (cleptric) [inline src/ORM/Association/BelongsToMany.php]: Same condition again?

File: src/ORM/Association/BelongsToMany.php [hunk 2]
```diff
@@ -783,7 +838,8 @@ protected function _saveLinks(EntityInterface $sourceEntity, $targetEntities, $o
         $belongsTo = $junction->association($target->getAlias());
         $foreignKey = (array)$this->getForeignKey();
         $assocForeignKey = (array)$belongsTo->getForeignKey();
-        $targetPrimaryKey = (array)$target->getPrimaryKey();
+        // $targetPrimaryKey = (array)$target->getPrimaryKey();
```
Comment (cleptric) [inline src/ORM/Association/BelongsToMany.php]: Accidentally left over?

Comment (inoas) [inline src/ORM/Association/BelongsToMany.php]: It is just a proposal.

File: src/ORM/Association/BelongsToMany.php (hunk 1 above)
Comment (inoas) [inline src/ORM/Association/BelongsToMany.php]: Copy-paste error. Let me fix it.

**Comment by inoas** [comment]
This code already runs as part of my patched core. If you think its good I will create test cases for it.

The issue https://github.com/cakephp/cakephp/issues/10288 describes why this is desireable.

In general you can set the foreignKey and bindingKey for all associations but only the targetBindingKey for belongsToMany was missing (targetForeignKey was already there). This brings BTM in line with the other associations.

File: src/ORM/Association/BelongsToMany.php [hunk 3]
```diff
@@ -102,6 +102,13 @@ class BelongsToMany extends Association
     protected $_saveStrategy = self::SAVE_REPLACE;
 
     /**
+     * The name of the field representing the binding key to the target table
+     *
+     * @var string|array
```
Comment (dereuromark) [inline src/ORM/Association/BelongsToMany.php]: |null

Comment (inoas) [inline src/ORM/Association/BelongsToMany.php]: Copied from  protected $_targetForeignKey; Patching

File: src/ORM/Association/BelongsToMany.php [hunk 4]
```diff
@@ -157,6 +164,52 @@ class BelongsToMany extends Association
...
+
+    /**
+     * Sets the name of the field representing the binding key to the target table.
+     * If no parameters are passed current field is returned
+     *
+     * @deprecated 3.4.3 Use setTargetBindingKey()/getTargetBindingKey() instead.
+     * @param string|null $key the key to be used to link both tables together
+     * @return array|string|$this
+     */
+    public function targetBindingKey($key = null)
```
Comment (robertpustulka) [inline src/ORM/Association/BelongsToMany.php]: Should we add and immediately deprecate method?

Comment (inoas) [inline src/ORM/Association/BelongsToMany.php]: I added it for consistency sake.

Comment (markstory) [inline src/ORM/Association/BelongsToMany.php]: Adding methods that are deprecated doesn't make sense to me. I would not include this method.

**Review by markstory — CHANGES_REQUESTED** [review]
We will need tests before this can be merged.

**Comment by lorenzo** [comment]
This is missing the ability to set the keying an array. Please add it to the constructor

**Comment by lorenzo** [comment]
It is also missing tests... are you also sure that was the only place where the primary key was used directly? If we are going to support this new option we need to make sure that it i used everywhere

File: src/ORM/Association/BelongsToMany.php (hunk 4 above)
Comment (inoas) [inline src/ORM/Association/BelongsToMany.php]: Will remove.
TLDR train of thought: people using the old interface maybe want consistent access when using this

**Comment by inoas** [comment]
@lorenzo this was just a proposal. It seems you are rather open to merge it once all issues are resolved and test cases have been added.

So I will look at a) where $target->getPrimaryKey() is being used and replace it with $this->getTargetBindingKey() and I will create tests. Okay?

**Comment by lorenzo** [comment]
@inoas sounds good :)

**Comment by inoas** [comment]
A quick look: As weak information the string `$target->getPrimaryKey()` exists throughout the framework within, **however these may not be the only cases to look at**:
- `Association::_addAssociationsToTypeMap()` (maybe wrong?)
- `Query::_addAssociationsToTypeMap()` (maybe wrong?)
- `HasMany::unlink()` (maybe wrong?)
- `ExistsIn::__invoke()` (looks 100% correct!)
- `BelongsToMany (will fix that)`
- `Marhsaller::_belongsToMany()` (will try to fix that!)

As I am not too familiar with the core and I don't want to create a massive PR, I will try to fix just BelongsToMany for now. Then I can take a look at `Marhsaller::_belongsToMany()`.

In reality all A) `->primaryKey();` and B) `->getPrimaryKey();` must be confirmed.
A quick search for all A) I would say the usage is always correct.

There are 25 total places where `->getPrimaryKey()` is being used as 3.4.2. So these need to be checked if the primaryKey is meant or the assoc binding (then bindingKey needs to be used).

**Comment by inoas** [comment]
Aside the uses in class BelongsToMany I will try to patch these:

#### Association
Method: `Cake\ORM\Association::_appendNotMatching($query, $options)`
Snip: `$primaryKey = $query->aliasFields((array)$target->getPrimaryKey(), $this->_name);`
Ref: https://github.com/cakephp/cakephp/blob/e91a091b269c35719cf7476d1bb1c8f3815308b6/src/ORM/Association.php#L956

#### Mashaller
Method: `Cake\ORM\Marshaller::_belongsToMany(Association $assoc, array $data, $options = [])`
Snip: `$primaryKey = array_flip((array)$target->getPrimaryKey());`
Ref: https://github.com/cakephp/cakephp/blob/28ea54b6a89c17dd9e30aa390e4d6b35a7c3daee/src/ORM/Marshaller.php#L367

Method: `Cake\ORM\Marshaller::_loadAssociatedByIds($assoc, $ids)`
Snip: `$primaryKey = (array)$target->getPrimaryKey();`
Ref: https://github.com/cakephp/cakephp/blob/28ea54b6a89c17dd9e30aa390e4d6b35a7c3daee/src/ORM/Marshaller.php#L457

#### Query
Method: `Cake\ORM\Query::_addAssociationsToTypeMap($table, $typeMap, $associations)`
Snip: `$primary = (array)$target->getPrimaryKey();`
Ref: https://github.com/cakephp/cakephp/blob/28ea54b6a89c17dd9e30aa390e4d6b35a7c3daee/src/ORM/Query.php#L414

**Comment by lorenzo** [comment]
> Method: Cake\ORM\Association::_appendNotMatching($query, $options)

Not needed, the method is overridden in BelongsToMany

> Method: Cake\ORM\Marshaller::_belongsToMany(Association $assoc, array $data, $options = [])

Yes, this will need to change

> Method: Cake\ORM\Marshaller::_loadAssociatedByIds($assoc, $ids)

Probably? I'm currently not sure about the need to change these, given that the feature is called `_ids` and you associate them directly by primary key.


> Method: Cake\ORM\Query::_addAssociationsToTypeMap($table, $typeMap, $associations)

No, you can leave this alone :)

**Comment by robertpustulka** [comment]
@inoas still interested in finishing this?

**Comment by inoas** [comment]
The only thing missing are a few unit tests, right?
I am very constrained at the moment with time, but if you can point me at where I would find similar unit tests I could adapt them to the double-sided custom foreign key setup here and then see if I can make them work...

**Comment by inoas** [comment]
This needs to be done:

>> Needs to change?:
>> Method: Cake\ORM\Marshaller::_belongsToMany(Association $assoc, array $data, $options = [])

> Yes, this will need to change

My expertise with the ORM internals is not sufficient to judge upon this:

>> Needs to change?:
> > Method: Cake\ORM\Marshaller::_loadAssociatedByIds($assoc, $ids)

> Probably? I'm currently not sure about the need to change these, given that the feature is called _ids and you associate them directly by primary key.

**Comment by dereuromark** [comment]
Do we still want to merge this into 3.x or 4.x?

**Comment by infokani** [comment]
> Do we still want to merge this into 3.x or 4.x?

Really need this feature

**Comment by othercorey** [comment]
This should be fixed in 4.1. 

https://github.com/cakephp/cakephp/pull/14377

**Comment by ADmad** [comment]
> Really need this feature

You can cleanup this patch and submit a new pull request. Simply waiting for someone else to do it for you won't help, as evident from this closed PR :)

**Comment by ADmad** [comment]
> This should be fixed in 4.1.

Ah well, guess someone else did implement it :)
